edd-discount-widget.php

```diff
@@ -135,20 +141,49 @@ function form( $instance ) {
 
 
         function get_discounts( $site_url, $api_key, $api_token, $max_discounts ) {
+
+            $site_url      = sanitize_text_field( $site_url );
+            $api_key       = sanitize_text_field( $api_key );
+            $api_token     = sanitize_text_field( $api_token );
+            $max_discounts = absint( $max_discounts );
+
+            if ( empty( $site_url ) || empty( $api_key ) || empty( $api_token ) ) {
+                return __( 'Widget settings are incomplete.', 'edd-discounts-widget' );
+            }
+
             $options = array(
                 'timeout'   => 5
             );
 
             $temp_url = parse_url( $site_url );
-            if( !$temp_url['scheme'] == 'http' && !$temp_url['scheme'] == 'https' )
-                $site_url = 'http://' . $site_url;
+            if ( isset( $temp_url['scheme'] ) ) {
+                if( !$temp_url['scheme'] == 'http' && !$temp_url['scheme'] == 'https' ) {
+                    $site_url = 'http://' . $site_url;
+                }
+            }
 
             $discounts = wp_remote_get( $site_url . '/edd-api/discounts?key=' . rawurlencode( $api_key ) . '&token=' . rawurlencode( $api_token ) . '&format=json', $options );
-            $discounts = $discounts['body'];
 
-            if( !$discounts || is_wp_error( $discounts ) ) $discounts = __( 'An unknown error occurred!', 'edd-discounts-widget' );
+            if ( is_wp_error( $discounts ) ) {
+                return esc_html( $discounts->get_error_message() );
+            }
+
+            $response_body = json_decode( wp_remote_retrieve_body( $discounts ), true );
+
+            if ( is_wp_error( $response_body ) ) {
+                return esc_html( $response_body->get_error_message() );
+            }
+
+            if ( isset( $response_body['error'] ) ) {
+                return esc_html( $response_body['error'] );
```

Comment:
The errors here do not need to be escaped, because you escape them when you actually display them. (So don't escape here, just escape on final display as late as possible.)